drivers/qemu/driver.go

```diff
@@ -0,0 +1,647 @@
+	return nil
+}
+
+func (d *QemuDriver) TaskConfigSchema() (*hclspec.Spec, error) {
+	return taskConfigSpec, nil
+}
+
+func (d *QemuDriver) Capabilities() (*drivers.Capabilities, error) {
+	return capabilities, nil
+}
+
+func (r *QemuDriver) Fingerprint(ctx context.Context) (<-chan *drivers.Fingerprint, error) {
+	ch := make(chan *drivers.Fingerprint)
+	go r.handleFingerprint(ctx, ch)
+	return ch, nil
+}
+
+func (d *QemuDriver) handleFingerprint(ctx context.Context, ch chan *drivers.Fingerprint) {
+	ticker := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-d.ctx.Done():
+			return
+		case <-ticker.C:
+			ticker.Reset(fingerprintPeriod)
+			ch <- d.buildFingerprint()
+		}
+	}
+}
+
+func (d *QemuDriver) buildFingerprint() *drivers.Fingerprint {
+	fingerprint := &drivers.Fingerprint{
+		Attributes:        map[string]string{},
+		Health:            drivers.HealthStateHealthy,
+		HealthDescription: "healthy",
+	}
+
+	bin := "qemu-system-x86_64"
+	if runtime.GOOS == "windows" {
+		// On windows, the "qemu-system-x86_64" command does not respond to the
+		// version flag.
+		bin = "qemu-img"
+	}
+	outBytes, err := exec.Command(bin, "--version").Output()
+	if err != nil {
+		// return no error, as it isn't an error to not find qemu, it just means we
+		// can't use it.
+		fingerprint.Health = drivers.HealthStateUndetected
+		fingerprint.HealthDescription = ""
+		return fingerprint
+	}
+	out := strings.TrimSpace(string(outBytes))
+
+	matches := reQemuVersion.FindStringSubmatch(out)
+	if len(matches) != 2 {
+		fingerprint.Health = drivers.HealthStateUndetected
+		fingerprint.HealthDescription = fmt.Sprintf("failed to parse qemu version from %v", out)
+		return fingerprint
+	}
+	currentQemuVersion := matches[1]
+	fingerprint.Attributes[qemuDriverAttr] = "1"
+	fingerprint.Attributes[qemuDriverVersionAttr] = currentQemuVersion
+	return fingerprint
+}
+
+func (d *QemuDriver) RecoverTask(handle *drivers.TaskHandle) error {
+	if handle == nil {
+		return fmt.Errorf("error: handle cannot be nil")
+	}
+
+	var taskState QemuTaskState
+	if err := handle.GetDriverState(&taskState); err != nil {
+		d.logger.Error("failed to decode taskConfig state from handle", "error", err, "task_id", handle.Config.ID)
+		return fmt.Errorf("failed to decode taskConfig state from handle: %v", err)
+	}
+
+	plugRC, err := utils.ReattachConfigToGoPlugin(taskState.ReattachConfig)
+	if err != nil {
+		d.logger.Error("failed to build ReattachConfig from taskConfig state", "error", err, "task_id", handle.Config.ID)
+		return fmt.Errorf("failed to build ReattachConfig from taskConfig state: %v", err)
+	}
+
+	pluginConfig := &plugin.ClientConfig{
+		Reattach: plugRC,
+	}
+
+	execImpl, pluginClient, err := utils.CreateExecutorWithConfig(pluginConfig, os.Stderr)
+	if err != nil {
+		d.logger.Error("failed to reattach to executor", "error", err, "task_id", handle.Config.ID)
+		return fmt.Errorf("failed to reattach to executor: %v", err)
```

Comment:
Thought for future refactoring: maybe we should create standard errors for this, reattach building, and driver state decoding and let shared code on the client handle logging them nicely.